### ISSUE TITLE
Decoder rejects long (but valid) '0' bits

### ISSUE THREAD

**Issue by Neil-McK** [littleyoda/sigrok-DCC-Protocoll#4]
I like this.  I have been working for some time on an Arduino-based DCC analyser (following on from work by RudiB) and have recently started using Sigrok.  I was surprised that there wasn't already a DCC decoder in Sigrok - I think this one should be submitted, even if it isn't yet complete.

I did encounter a problem when sampled the output from DCC-ex controller.  Most controllers including DCC++ generate DCC signals with a 58us half-period for a '1' and 100us half-period for a '0', in line with the NMRA standard S-9.1.  However, the standard is very tight about the '1' bit generation but not the '0' bit, so a '0' bit may be significantly longer (between 90us and 10000us for either or both half-periods - parameter tr0 - with a maximum total period of 12000us - parameter t0total).  DCC-ex capitalises on this to simplify its signal generation, by creating 58us half-period '1' bits and 116us half-period '0' bits; thus a single regular clock of 58us generates both the '1' and '0' bit signals.

However, the Sigrok DCC decoder imposes constraints on the acceptance of '0' bits which aren't (as far as I can see) in the NMRA standards, and will treat a bit with a half-period of longer than 116us as invalid.  Consequently, when the capture rate is high (>1MHz) to get accurate measurements, I observe that some valid '0' bits are rejected by the decoder.  At a lower capture rate (~50kHz), the decoder relaxes the required accuracy and accepts all sampled bits as valid.

Can you consider adjusting the criteria for the '0' bits in the decoder so that it is in line with the standard?

Note:  The standard isn't clear about why asymmetric (stretched) '0' bits are permitted.  Is it to allow the average voltage to be driven to non-zero, permitting a controller to drive non-DCC (DC) engines alongside DCC ones?

**Comment by Neil-McK** [comment]
Correction: I've just checked and the decoder treats bits with half-period of longer than **119**us as invalid (although the associated comment says **116**us).  I thought that I must have changed it to 119 but that's how it is in GitHub.  However, the comment still stands that the decoder is rejecting '0' bits longer than 119us that appear, from the standard, to be valid.

**Comment by FrightRisk** [comment]
As you can see from my commit, I've been using this for a while.  I'm trying to see what Noell changed and making sure the accessory addresses are still correct. The zero stretching was and is a bad idea. I don't believe it is in the NMRA spec, just that they "allow" it for people to be able to so thinks like we are with DCC-EX and for operating a DC loco. It makes DCC have a DC component by putting one side of the pulse have longer zeros. So one half of the track has a shift that lets a motor see DC. Anyway, bad, bad, bad ;) 

Looking at the code it appears they are trying to handle the zero bit stretching. I'll have to see how accuracy is affecting this. I noticed that in theory, I should have been able in the old version to sample at any rate over 24kHz and have it work. Certainly at 50. But it didn't, I had to sample at 100kHz to get accurate packet readings. It will be interesting to try this version.

https://github.com/littleyoda/sigrok-DCC-Protocoll/blob/7b2b00f1e3fa8e813033cbfdea34777a6da8a54b/dcc/pd.py#L1303

**Comment by FrightRisk** [comment]
Hi Neil, are you on our DCC-EX Discord? There is another XXX-Mck but it's not Neil. I could use help fleshing out things like this for the website. Coincidentally, I was working on a page for this for a couple of days. https://dcc-ex.com/reference/tools/diagnostic-tools.html

**Comment by Neil-McK** [comment]
Bear in mind also that there's about 8us of jitter on the DCC++ ex signal as well (probably caused by clock interrupts in the Arduino that occur every 1ms), so 0 bits are 108-124us in length and 1 bits are 50-66us, before allowing for measurement inaccuracies.  It's debatable whether this should be treated as an error by the analyser, I think it should not.  I agree that an asymmetric signal is bad and it's worth annotating them.   But the fact is that a valid NMRA zero bit may be up to 10000us, so failing to decode a packet containing 0 bits that are well within this seems undesirable in an analyser.

I would suggest that undesirable or unexpected signal behaviour be annotated in a warning area, but not cause a resynchronisation of the decoder unless it falls securely outside of the NMRA spec.

I'm not on discord, I'll look at it after Christmas is over!

**Comment by Neil-McK** [comment]
I've just checked the Atmega datasheet and the Timer1 and Int0/Int1 interrupts are all higher priority than the Timer0 interrupt used by the Arduino clock, so it looks superficially as if Timer0 shouldn't affect the DCC++ ex waveform.  However, once a Timer0 interrupt has been detected by the CPU, higher priority interrupts are blocked until the Timer0 interrupt code finishes and executes RETI (or executes SEI).

**Comment by FrightRisk** [comment]
And this jitter is NOT on DCC++ Classic that uses 2 timers? Or is it? Merry
Christmas!

Fred

**Comment by FrightRisk** [comment]
Forgot one more thing, the datasheet says you must have a .1uf cap across
the Vcc and Gnd, I wonder if you tried that in your tests. I see a lot of
people skip that step.

**Comment by Neil-McK** [comment]
I tried a ceramic one but it made no difference.

I haven't checked DCC++ classic, it may be worse, but half of the wave is
generated by the timer hardware rather than interrupts, so we'll see.

Hope you had as good a Christmas as possible in the current environment!

**Comment by FrightRisk** [comment]
Were your  tests on PROG or MAIN? I went back and we saw that some code
that was supposed to stabilize the wave form was missing for some reason.
So we put it back. That is in the master branch now. It will go in the next
release.

Fred

**Comment by Neil-McK** [comment]
My tests were all on Main - I haven't even looked at the Prog signal.

I've tried the original DCC++ classic on a Mega.  Using Sigrok I 
couldn't see any pulses that were more than a microsecond away from 58 
or 108 us.  And I modified my sniffer so that it disabled Timer0 (used 
Timer1 without interrupts to get microsecond and millisecond counts).  
After further modifying it so that the Serial USB handling didn't 
disrupt the measurements, I now get upper and lower bounds for pulse 
length of 57-59 us and 107-109 us.  For DCC-ex (based on release 3.0) I 
get 51-65 us and 110-123 us from exactly the same test rig, i.e. a 
spread of around 6-7 us either way from the nominal pulse length.

I've just downloaded the current master version (as ZIP) from github, 
compiled and loaded it.  I still get 51-65 us and 109-123 us pulse 
lengths.

In the original DCC++ classic, the DCC waveform wasn't dependent on 
timely execution of the interrupt code, as the waveform was generated by 
the timer hardware.  Only the period of the next cycle was modified by 
the interrupt routine.  However, the DCC ex code, while being very 
elegant in its use of only one timer, generates the waveform from within 
the interrupt response code so is more sensitive to other interrupts.

It's rather academic though, as most of the DCC frames are generated 
without disruption (only a small percentage is affected by coincident 
interrupts) so it's not likely to affect train operations unless a 
decoder is particularly sensitive to the pulse length and throws a 
wobbly.  Hence my suggestion that the Sigrok DCC decoder and the MynaBay 
DCC Arduino Library might be a bit more forgiving of the measured bit 
length.

NB I've done some experiments regarding the impact of the Timer0 
interrupt on loop code and on interrupt handlers, and will post them on 
my http://wakwak2popo.wordpress.com in a day or two.

Neil.

PS I have just noticed that I'd tweaked the classic software's DCC pulse length from 100us to 108us.  However, the gist of the message still stands that the figures are dead stable with the classic software.

------ Original Message ------

**Comment by FrightRisk** [comment]
Neil, I sent you a message on Trainboard yesterday since I didn't have
another private email. Could you check there? Thanks!

Fred

**Comment by FrightRisk** [comment]
And apologies, the changes were not merged into master yet, they were in
the anti-jitter branch. I merged them now.

**Comment by su68es** [comment]
Hello,
a short explanation about my version of the decoder.
The goal was to decode a DCC signal as good as possible, independent of the quality or the temporal correctness of the signal.
However, the higher the sampling rate, the more accurate the signal must be.
To check the quality, the "timing" function of sigrok should be used in parallel with the DCC decoder and there should be a resolution of at least 1µs (i.e. sampling rate at least 1MHz).
Many greetings
Roland

**Comment by su68es** [comment]
In fact, the comment is wrong. It must be "90µs - 119µs".
The values are from the standard "http://normen.railcommunity.de/RCN-210.pdf" table page 4 - rightmost column (values valid at the input of a decoder).
I will adjust the comment on occasion.

**Comment by su68es** [comment]
Here is a short explanation of the sampling rate and DCC detection:
At a sampling rate of 25kHz the state in the DCC data stream is measured every 40µs. This is sufficient to detect all positive and negative half-waves in the data stream. However, with DCC one bit always consists of one positive and one negative half-wave. To detect DCC, the length of the respective half-wave must be measured (as accurately as possible). A one-bit consists of two half-waves of 58µs +/- 3µs tolerance each. But if now only every 40µs a measurement is made, the length of the half-wave can only be determined very inaccurately.
Therefore the DCC evaluation in the sigrok-decoder allows an additional deviation of 80µs at 25kHz (2 x 40µs).
"accuracy = 1/self.samplerate*1000000 #µs".
With this an evaluation of the signal works if it is reasonably good. A statement about the quality does not work at 25kHz.
The higher the sampling rate, the less is the additional allowed deviation. At 1MHz a deviation of 2µs is still allowed.
Since DCC considers tolerances in the range of 3µS, from my point of view the sampling rate must be at least 5MHz to judge the timing of DCC. The additional deviation is then only 0.4 µs.

**Comment by su68es** [comment]
To return to the initial situation. From my point of view, a 0 bit half must not be longer than 119µs or 10000µs. So the implementation does not have to be adjusted and the point can be closed?

**Comment by FrightRisk** [comment]
Maybe. If my logic is correct, the Nyquist theorem says that you can sample at twice the highest frequency for analog to digital conversion. I am not sure that applies to digital pulses. In any case, with the last version of this decoder, twice 58us would be sampling at 34kHz, so I tried 54kHz or whatever the dropdown selector had for something in the 50s range. It would not decode correctly. I had to go to 100kHz and it worked just fine. I will try again with DCC-EX and different sampling rates

**Comment by su68es** [comment]
Hello Fred,
I wish you a happy and healthy new year.
The Nyquist theorem also works in the DCC decoder. But the point is that thus all edge changes are recognized correctly, but thus the DCC signal cannot be evaluated.
For the DCC decoding not the state changes are relevant but the varying the duration of the parts of the bits.
I have attached a graphic.
![image](https://user-images.githubusercontent.com/76778984/103438867-32420880-4c38-11eb-8e0b-9b6b27f7b31b.png)

Assumed the DCC bit sequence 1-0-1-0 is to be generated and decoded.
Line 1 shows the signal to be generated (1-0-1-0-1-0-1-0).
Line 2 shows the corresponding timing.
Line 3 shows the trigger points of a 25kHz sampling.
Line 4 shows that the sequence 1-0-1-0-1-0-1-0-1 is reliably detected (Nyquist theorem).
In lines 5 and 6 you can see what is generated in the DCC decoder from the sampling result.
Edge changes in a 40µs grid. This does not allow to decode the DCC signal reliably.
Therefore: The higher the sampling rate, the more accurate the length measurement and thus the signal evaluation.
My Tams command station e.g. generates a signal which can be decoded even at 25kHz.

25kHz Sampling
![image](https://user-images.githubusercontent.com/76778984/103438891-6ae1e200-4c38-11eb-93c9-e2dba18c93a4.png)

24MHz Sampling
![image](https://user-images.githubusercontent.com/76778984/103438908-8cdb6480-4c38-11eb-9f2e-8d2b0cdcdf13.png)

At 50kHz it should actually already work if the generated signal is good.
If you want you can send me two files. One with 50kHz sampling where it doesn't work and the same signal in a high resolution. I will have a look at it then.

Greetings
Roland

**Comment by ghost** [comment]
Can I jump in there? I'm totally new to DCC, so I might be wrong, but to my understanding the following situation should be allowed for representing zero by the specs and currently is interpreted as "dirty signal" by the decoder: Low 150us, High 150us = 0 bit.

I would refer to the following NMRA standard: https://www.nmra.org/sites/default/files/s-91-2004-07.pdf

Which says (highlighting by me):

> In a "0" bit, the duration of the first and last parts of each transition shall nominally be **greater than or equal to 100 microseconds**.  To keep the DC component of the total signal at zero as with the "1" bits, the first and last part of the "0" bit **are normally equal to one another**. Digital Command Station components shall transmit "0" bits with each part of the bit having a duration of between 95 and 9900 microseconds with the **total bit duration of the "0" bit not exceeding 12000 microseconds**.  A Digital Decoder **must accept bits whose first or last parts have a duration of between 90 and 10000 microseconds as a valid bit with the value of "0"**.  Figure 1 provides an example of bits encoded using this technique.

I would interpret the "or" in the last highlighted section as a non-exclusive OR (otherwise they should have written "either or"), especially because previously they say that usually both parts should have equal lengths. Thus, a signal with e.g. 6000us low, 6000us high should be considered a valid 0 bit.

In the current state [the code expects either](https://github.com/littleyoda/sigrok-DCC-Protocoll/blob/7b2b00f1e3fa8e813033cbfdea34777a6da8a54b/dcc/pd.py#L1309):

- part 1 to be long, part 2 to be short
- part 1 to be short, part 2 to be long

The code currently does handle both parts being long as invalid.

In my situation I was playing around with rocrail and a USB-to-serial. I do get some 0 represented by 150us low/high combinations (300us total). I cannot rule out that something is wrong with my setup, but from reading the specs I would assume that this should be a valid 0 representation.

I was able to get the code working for my situation with the check `elif 90-accuracy <= part1 <= 10000+accuracy and 90-accuracy <= part2 <= 10000+accuracy` but cannot say if this might have undesired side-effects.

After this change, the interpreted rocrail commands look quite OK with some train speed commans being decoded.

Update: Added a screenshot.

![2021-01-01-200156_1880x683_scrot](https://user-images.githubusercontent.com/3071022/103444777-5d464f80-4c6c-11eb-8808-4f827191d9a7.png)

**Comment by FrightRisk** [comment]
@su68es Hi Roland. And thanks for the greetings. What a great post also!! A Picture is worth a thousand words. Thank you for taking the time to include all those. I only have one logic analyzer, a cheap Saleai, but I can try a few some different decoders I built and run software on an arduino to double-check my signal quality. I can compare DCC-EX with Classic. We are also experimenting on a branch to do our timing differently yet again. The old DCC++ used 2 timers. We use one. But we thought of some other ways to deal with interrupts and timing. I did not want to lose the major benefits we have now, which is the ability to put the signals on any pin to match motor boards, and not needing jumper wires. But we could have a config file option for "more accurate timing". I agree with what you said in your post. Happy New Year!

**Comment by FrightRisk** [comment]
@aufziehvogel Grüß Gott Stefan (oder Guten Tag ;) , are you a Murakami fan or just like mechanical birds? ;) Well, you have run into what happens with a loosely organized and political "standards" committee ;) We have struggled with this in designing and supporting DCC++ EX. Without too harsh a judgement, I would say that the NMRA DCC standard is difficult to understand and thus not well implemented. It reminds me of an old "Saturday Night Live" sketch where the head nuclear engineer is retiring and going on holiday to a tropical island. His last bit of advice to his team was, "Just remember, you can NEVER push the cooling rods in too far!". Only after he was unable to be contacted did the debate rage as to whether that warning was meant as a safety mechanism as in, "there is no such thing as pushing the cooling rods in to far because better pushed in far than not!" or as a warning of danger as in, "whatever you do, to not push the rods in too far or there could be major problems".  The sketch ended with the pensioner on a beach with a fruity cocktail looking out into the ocean and seeing a mushroom cloud rising above the horizon. ;) DCC is a bit like that. lol. I could share some of how things got this way, but the two biggest reasons are manufacturers with some clout forming some of the way addresses work and not planning far enough into the future as to how DCC would be used. Just look at 24 bit, 48bit and 128bit addressing and how the packets look ;) Same with "zero bit stretching" which is the real subject of my reply.

Before I go into zero stretching, I will mention that our biggest support nightmare other than WiFi, are all the issues with decoders, especially when trying to read and write CVs. I don't know how much money is to be made in this hobby, but I believe a lot decoder companies are small. They read the NMRA specification just like you and I and make their own determinations. Or they assume you will use their command station.  I can tell you that whatever decoder you have, if it is truly generating 150us pulses, is clearly out of spec. It would be interesting to see if its ACK pulses were the proper duration and level. That is an ongoing problem we run into every day. Adding further to confusion is all the places and different versions of the standard. Here is the correct document which is different: https://www.nmra.org/sites/default/files/standards/sandrp/pdf/s-9.1_electrical_standards_2020.pdf

Zero bit stretching is a bad idea. It is a remnant left over from 20 years ago for people who wanted to run ONE DC loco on a DCC track. Without going into all the reasons why this was a kludge that should have been avoided, I will stay that as with the nuclear engineers, you are just misreading the spec. The ONLY reason why you would have a 0 pulse of more than 100us (give or take a tolerance factor) is to allow a DC loco to run on a DCC track with zero bit stretching. If you look closely at the wording again, you will see one of the sentences you partially highlighted;

"To keep the DC component of the total signal at zero as with the "1" bits, the first and last part of the "0" bit are normally equal to one another". 

The spec mentions "DC Component" (as opposed to DCC), and relates to the fact that DCC is a bipolar square wave with no ground reference. One track is fully energized, then goes to zero and then the other is energized. If you put a DC loco or voltmeter on the track it would see zero volts because the alternating cycles cancel each other out. There is what appears like an AC square wave on the track, but the DC component is zero. If you stretch the zeros, BUT ONLY ON EITHER THE UPPER OR LOWER PULSE BUT NOT BOTH, you are no longer in equilibrium. The voltage spends more time on one rail than the other. If you increase the duration of rising part of the pulse compared to the falling pulses, DCC devices could still decode the pulses, the alternating waveform could still be rectified into DC to control the decoder and the loco motor, yet we have introduced a "DC component" to the tracks. A DC motor will see more current trying to turn it in one direction than the other. You can control the speed by controlling how much longer the rising pulse is than the negative one. Direction is controlled by switching the longer pulse to the other side of the waveform and thus reversing the DC component of the voltage.

We know both parts of 1 bits are ALWAYS the same duration. And here is where the spec here is confusing as usual. Ideally, one part of the pulse could vary, but the other would always be 100us. I can't think of a reason you would ever want the other side of the pulse greater than that. It makes "zero" sense ;) and may destroy a DC motor in addition to it taking away from the DC bias you are trying to put on the rails with the other side of the pulse! So zero bits in ONE direction can be longer up to a maximum of 9900us and a total duration of 12000us for both parts. That could be read as a 9900us rising pulse, then the maximum falling be 2100us. So while technically, it would seem a 9900 up and 2100 down wave should be read as a 0, it does not seem correct. But I'll ask around. It SHOULD be an exclusive or, if zero bit stretching is not turned on to run a DC loco, there is no reason why the pulses should not be equal and in spec.

Sorry for rambling, but to put it in other terms:

1. A 1 bit ideally is 58us, NMRA will allow you a tolerance for sending a one bit from 55 to 61us
2. A decoder needs to read a 1 as anything from 52 to 64us just to make sure we get it
3. A zero ideally is 100us, NMRA for pure DCC would allow from 95 to 110us, but if you want to turn on zero stretching for a DC loco, will allow one or the other but not both sides of the same pulse to be up to 9900us. with both sides not exceeding 12000us
4. A decoder should decode a 0 bit where either part of the pulse is from 90 to 10000us, where the tolerance for reading the other side of the pulse should be 110.
5. I have no idea at the moment what the 12000us has to do with anything, but will research it.

Best Regards,

Fred
DCC-EX Team

**Comment by FrightRisk** [comment]
@aufziehvogel I forgot to ask, what Command Station is this? What is generating the DCC signal?

**Comment by su68es** [comment]
Hello all,

@ Fred: Glad we agree on the timing. I will add to the description as follows:
Decoding of DCC is done by measuring the time interval of the zero crossings (edge changes). Therefore a high sampling rate is necessary for an accurate measurement. At sampling rates below 1MHz the decoding may be inaccurate depending on the signal quality.

@ Stefan: I fully agree with Fred's explanations. The specification is not clear here, but extending both half-waves makes no sense. Also from my point of view a signal 150µs/150s or 6000µs/6000µs is invalid.
I will make the signal detection for streched-zero switchable by configuration, because the streched-zero is not needed today and could also still collide with Railcom. So the received signal can then be checked 'better'.
Like this:
![image](https://user-images.githubusercontent.com/76778984/103455091-3b8bad80-4cea-11eb-83b2-7692f6c2b737.png)

@ Fred: I think with the streched-zero identifier turned off, both half-waves are allowed between 90µs and 119µs each. The Railcommunity standard is a bit clearer here. The 119µs is 116µs +3µs tolerance (116 µs to maintain compatibility with other protocols and protocol extensions).

Kind regards
Roland

**Comment by Neil-McK** [comment]
Fred,

We should bear in mind the context of this discussion - we're talking 
about a diagnostic tool which is aimed at  helping to understand what's 
going on in the DCC signal.  The current implementation does this until 
it finds a pulse whose length is outside of its stringent requirements, 
then it throws a wobbly and does not decode until the next sync sequence 
is detected.  It would be better if the decoder continues to decode 
while it can.  As it stands, it would be difficult to correlate 
occurrences of long bits against packet type, for example.

On the subject of the NMRA standard, it is a standard, no more and no 
less.  It does not present a full specification of the DCC signal or a 
complete design for a decoder or command station.   It only prescribes 
those aspects which are REQUIRED for conformance, in order that diverse 
equipment may interoperate successfully.   One has to avoid 'filling in 
the gaps' by assuming things that are not specified.

Therefore, it specifies the minimum and maximum bit lengths that a 
command station must generate.  It specifies the minimum and maximum bit 
lengths that a decoder must be capable of detecting.  It does not 
however specify whether a decoder should accept or reject bit lengths 
outside that range, as that is not required to guarantee 
interoperability.  Consequently, a decoder that accepts and decodes '1' 
bits for pulses between 30us and 80us and '0' bits between 81us and 
10000us is NMRA compliant, but a command station that generates '1' bits 
as pulses of 54us  and '0' bits as pulses of 94us is not.  Likewise, a 
decoder that is too strict and only accepts '1' pulse lengths between 55 
and 61us would be non-compliant.

Regarding the contents of the specs, I don't know where you got a figure 
of 110us for upper bound on '0' pulses, it's not in the 2006 NMRA spec 
or the 2020 version.

S-9.1-2006 says:
"In a "0" bit, the duration of the first and last parts of each 
transition shall nominally be greater than or equal to 100 microseconds. 
To keep the DC component of the total signal at zero as with the "1" 
bits, the first and last part of the "0" bit are normally equal to one 
another. Digital Command Station components shall transmit "0" bits with 
each part of the bit having a duration of between 95 and 9900 
microseconds with the total bit duration of the "0" bit not exceeding 
12000 microseconds. A Digital Decoder must accept bits, whose first or 
last parts have a duration between 90 and 10,000 microseconds, as a 
valid bit with the value of "0".   Digital Decoders must accept one bits 
whose positive and negative components differ by as much as 6 
microseconds."
[ There is no requirement to reject or ignore bits that fall outside of 
this specification, and no other requirements regarding '0' bits. ]

S-9.1-2020 says much the same thing:
"In a “0” bit, the duration of the first and last parts of each 
transition shall nominally be greater than or equal to t0 (nominally 
100us). To keep the DC component of the total signal at zero as with the 
“1” bits, the first and last part of the “0” bit are normally of equal 
duration. Digital Command Station components shall transmit “0” bits 
with each part of the bit having a duration within the t0 range (95 to 
9900us) with the total bit duration of the “0” bit not exceeding t0tot 
(12000us).  A Digital Decoder must accept bits, whose first or last 
parts have a duration within the tr0 range (90-10000us) as a valid bit 
with the value of “0”.  "

Consequently, any decoder that only detects '0' bits if the pulse length 
is between 90 and 119 us is not compliant with the NMRA standard.  And a 
command station that, for design or implementation reasons, generates 
'0' pulses of 130us, 150us or 116us (rather than 100us), is 
NMRA-compliant.

Lastly, you seem to be forgetting that DCC++ EX generates '0' pulses 
with a nominal half-length of 116us which is NMRA-compliant but outside 
of your desired decoder bit length of 90-110us.   Stefan's observed bit 
length of 150us is also NMRA-compliant.  It would be interesting to know 
how many other NMRA-compliant command stations generate '0' bits longer 
than 100us.

Regards,
Neil.

P.S.  I deliberately haven't mentioned 'stretched bits' at all above, 
and the parts of the NMRA spec quoted apply even when there is no DC 
component to the signal from asymmetric stretched bits.

------ Original Message ------

**Comment by su68es** [comment]
Hallo Niel,

the http://normen.railcommunity.de/RCN-210.pdf is quite clear on this point.
In a one bit, the first and the second half must have a duration of 58 µs ± 3 µs
in a one bit.
The duration of a one bit is therefore 116 µs ± 6 µs.
Both halves of the bit must have the same duration with a tolerance of ± 3 µs.
**A decoder must accept a tolerance for one bit half of ± 6 µs.
This means that a decoder must recognize such received bits as valid one bits,
whose two parts each have a duration between 52 µs and 64 µs.**

**In a zero bit, the nominal duration of both halves is 100 µs.
The minimum duration of half of a zero bit is 95 µs.
The maximum duration of the half of a zero bit is**
9900 µs if an analog locomotive is also to be controlled.
**116 µs to maintain compatibility with other protocols and
protocol extensions.**
The total duration of a zero bit must not exceed 12000 µs.
To keep the DC components of the complete DCC track signal at zero as with the
one bits to zero, both parts of the zero bit are normally of equal length.
In order to obtain a DC component for alternative control purposes, also only
half of a zero bit may be extended.
A decoder must recognize such received bits as valid zero bits, whose first or
second part has a duration between 90 µs and 10000 µs. 

Kind regards
Roland

**Comment by ghost** [comment]
@FrightRisk Hi Fred, thanks for the reply. I can understand what you mean. So the whole reason for long pulses being allowed is the DC hack, and thus normal 0 bits should always use ~100us signal parts.

My setup is rocrail with "dcc232" selected as controller, a USB to serial converter and a saleae clone logic analyzer. So, I would say rocrail acts as a command station in this case and sends the output directly to the USB-to-serial.

But I also read on their info page now that this mode [does not fully conform to NMRA](https://wiki.rocrail.net/doku.php?id=dcc232-en). And trying to send my own DCC signals over USB-to-serial I now understand that the USB-to-serial poses some limitations which make standard-compliant operation difficult.

**Comment by Neil-McK** [comment]
Hi Roland,

Yes, that's very much in line with the NMRA standard, the operative 
parts for decoders being
"A decoder must recognize such received bits as valid zero bits, whose 
first or
second part has a duration between 90 µs and 10000 µs."

and
"A decoder must recognize such received bits as valid one bits,
whose two parts each have a duration between 52 µs and 64 µs."

[The other bits (part of which you have bolded) relate to signal generation and
are not relevant to decoders.]

Also, the decoder has to take into consideration any potential measurement 
errors, so if the measurement has up to 4 µs error, then the decoder must accept
measured values of between 48 µs and 68 µs.

Of course, a decoder is not prohibited by either standard from 
recognising a wider range as valid zero or one bits, as this does not 
restrict its ability to interoperate with compliant command stations.  
Conversely, placing unnecessary restrictions on the bit lengths 
recognised by a decoder will prevent or restrict interoperation with 
command stations that are NMRA-compliant, and make the decoder or 
command station appear to be unreliable or broken.

The bit about a command station generating a maximum of "116 µs to 
maintain compatibility with other protocols and
protocol extensions" is nothing to do with the NMRA standard for DCC and 
probably relates to proprietary software and hardware.  In any case, 
this isn't relevant to the protocol analyser which is aimed squarely at 
NMRA standards.

Regards.

------ Original Message ------

**Comment by Neil-McK** [comment]
Stefan,

The reason for asymmetric long pulses is to generate a DC bias which 
supports analogue engines, but it potentially burns out the motor 
(because of the continuous but alternating high voltage) and so is 
discouraged.  However, the DCC standard also permits symmetric long 
pulses for '0' bits.  There is no requirement for '0' bits to be 
generated as 100 us, the original 2006 spec required 'greater than or 
equal to 100us'.  There are probably many reasons for wanting to 
generate symmetric pulses longer than 100us for '0' bits, the most 
obvious one being to simplify the hardware or software in the command 
station.

A case in point is DCC++ EX which, for convenience, generates '0' bits 
which are 116 µs, i.e. exactly twice the length of the '1' bit.  This permits a 
single 58us timer interrupt to be used to generate as many independent 
DCC signals as required (2 in fact).  For '1' bits, the digital output 
is toggled on consecutive interrupts, and for '0' bits, it is toggled on 
alternate interrupts.  The down side of a longer '0' bit is that fewer 
bits can be transmitted per second, and also that there is a risk of 
incompatibilities with decoders that are more stringent than the NMRA
spec when validating of pulse lengths.

I'm not sure why RocRail should use 150us pulses, but they must have a 
good reason, and it is in the valid range within the NMRA spec (both 
2006 and 2020 versions).  I suspect they have been pressured to add a 
warning because non NMRA-compliant decoders are rejecting valid '0' 
pulses of 150us in length, and no-one has investigated where the fault 
lies!  Hopefully, more and better diagnostic tools will make such 
analysis easier.

Regards.

------ Original Message ------

**Comment by su68es** [comment]
Hi Neil,

actually we want all the same. Check whether a generated signal meets the **minimum** requirements of a decoder. And these minimum requirements are unfortunately not clear.
It starts with the fact that I have based the software on the standard of the [RAILCOMMUNITY](http://www.vhdm.de/index.php?option=com_content&view=article&id=49&Itemid=61) not the NMRA. The RAILCOMMUNITY standard clearly states that a half of a "0" bit may only be greater than 119µs if an analog locomotive is to be controlled as well, is normally 100µs, and for compatibility with other protocols and protocol extensions up to 119µs are allowed. In the NMRA it is defined differently.
I think if I make it configurable both cases are covered:
1: half of a "0" bit maximum 119µs
2: half of a "0"-bit maximum 10000µs
For the 2nd case it still have to be defined if this is valid for both halves or if only one half may be greater than 119µs (current implementation). Or should I make this configurable as well?
What do you think?

Kind regards
Roland

**Comment by Neil-McK** [comment]
My main issue is how the decoder handles a suspect pulse (i.e. one whose
purpose is clear but it's outside the desired or accepted DCC lengths).
Currently the decoder puts a warning out and forces a resync, so I'm forced
to decode the affected frame by hand.  I would like to see it decode such
frames, but still mark them with a warning.  I realise that this is
somewhat more involved than changing just the pulse lengths, but with your
experience with the decoder it should be a doddle!

I like your suggestion of allowing configuration of the maximum '0' bit as
opinions vary on what is acceptable.  If the target output is 100us then
NMRA requires 90us to be accepted, so how about allowing 10us margin the
other side.  Using the RailCommunity target that the command station should
be trying to produce 100us or 116us (or something between) then a 'normal'
acceptable range for the decoder would be between 90 and 126us, anything
outside this producing a warning.  The option of a warning if there is a
significant imbalance between bit halves (a DC bias) would also be helpful.

Neil

**Comment by FrightRisk** [comment]
Great discussion! Sounds like some our Discord discussions about all the issues with the "stamdards". I also like the new option. Great idea Roland.

"116 µs to maintain compatibility with other protocols and
protocol extensions."

Do we have any examples of any of these protocols a d extensions? Rock rail? DCC-EX, but we only did that because we thought 116us was within spec. This whole thing made me think of embedding information in the zero bits. You could have 200 and 300us 0s for dots and dashes and send Morse code ;) seriously  though, I am going to imagine how this might be used. 

"The total duration of a zero bit must not exceed 12000 µs."

Where does this come from? It must have a reason.

Neil, thanks for your usual thorough logic. Sounds legal ;) in that vein, I would add that in law, like constitutional law, a judge looks at the intent of the law. They go back to writings of the authors. So knowing this was created to handle a DC loco is a start. And you make the point that a CS has rules about what it can generate, but that doesn't limit a decoder. What I want to know is the intent. Why would we allow stretched zeros without a DC loco. And why would it allow both sides of the wave to be stretched? Or is this a mistake in the spec?

"In order to obtain a DC component for alternative control purposes, also only
half of a zero bit may be extended.
A decoder must recognize such received bits as valid zero bits, whose first or
second part has a duration between 90 µs and 10000 µs"

This I agree with. So my question is the short part of the 0. What circumstance would warrant it being longer than a non DC pulse? Why would you want a 0 with each part being 6000us?  And where did 12000 come from? Why not 20000 for 2 times 10000 if they intended both sides to be stretched?

Fred

**Comment by Neil-McK** [comment]
Roland, Fred,

I've looked at the NMRA figures to try to understand how they would be derived, particularly the tolerance on the pulse lengths sent from a command station to be NMRA compliant, and the tolerance that a decoder must allow for reception.  It is evident that the tolerances are set at a maximum of 5% for the command station, and that the decoder must allow at least 10% tolerance for received bits:

NMRA Design target values:    
  0 bit:            >= 100us
  1 bit:            58us
  1 bit delta:   0us  (i.e. both halves are designed to be exactly 58us).
 
Acceptable values generated by an NMRA compliant command station, calculated as target +/- 5% tolerance:
  0 bit:            >=95us
  1 bit:            55us to 61us
  1 bit delta:   3us

Range that an NMRA compliant decoder must accept, calculated as target +/- 10% tolerance:
  0 bit:            >=90us
  1 bit:            52us to 64us
  1 bit delta:   6us

If we revise this table to include a maximum design target value of 116us as included in the Rail Community standard RCN-210 (2019), we get:

RCN-210 Design target values:    
  0 bit:            100us **to 116us**
  1 bit:            58us
  1 bit delta:   0us

Acceptable values generated by an RCN-210 compliant control station, calculated as target +/- 5%:
  0 bit:            95us **to 122us**
  1 bit:            55us to 61us
  1 bit delta:   3us

Range that an RCN-210 compliant decoder must accept, calculated as target +/- 10%:
  0 bit:            90us **to 128us**
  1 bit:            52us to 64us
  1 bit delta:   6us
 
So I suggest four selectable modes for the Sigrok DCC decoder:

1) For NMRA compliance testing of controllers (which should only be allowed at sample rates >=2MHz where the measurements are sufficiently accurate), use ranges 55-61 and 95-9900 us.  Suggest this mode could be NMRA-STRICT.

2) For NMRA decoding, use the Decoder ranges which are 52-64 and 90-10000 us. Suggest this be NMRA-FULLRANGE.

3) For RCN-210 compliance testing (which should only be allowed at sample rates >=2MHz where the measurements are sufficiently accurate), use ranges 55-61 and 95-122 us.  Suggest this be RCN210-STRICT.

4) For RCN-210 decoding, use the decoder ranges which are 52-64 and 90-128 us.   This could be RCN210-FULLRANGE

I think that would cover everything - do you agree?  

You will need to retain the existing calculation of the measurement tolerance on the above values of course, as the measurements get increasingly inaccurate as the sample rate decreases.

[EDIT: I've corrected a few typos in the original]

**Comment by Neil-McK** [comment]
Fred,

You may have been joking when you suggested "embedding information in the zero bits", but that's rather the point.   You can validly vary the '0' bit length and remain NMRA compliant, and decoders that strictly adhere to the NMRA standard will continue to work properly because they don't mind that the bit length varies.

The standard defines those aspects of a design and implementation that must be there in order to allow a controller to work with a decoder, irrespective of how the controller and decoder are implemented.

The standard has to be able to sustain future enhancements, e.g. new addressing modes, new frame types for new functions.  If a decoder is implemented according to the original standard then it will be able to continue to work with a control station implemented to the later standard, although ignoring the new features.  Likewise, a decoder that supports the new features will be able to work with an older control station.

Difficulties and incompatibilities come when the standard isn't adhered to, for example when a decoder manufacturer might think that all '0' bits should be nominally 100us +/- 10%, i.e. 90-110us and reject anything outside this range.  Then the decoder fails to work with DCC++ EX, but works with a LENZ command station or a Hornby command station.  

Who's to blame in this case?  Which component is non-compliant?  Many would argue DCC++ EX as it's using a different pulse length.  But 116us pulse length generation for '0' bits is valid according to the NMRA and RCN specs, and DCC++ EX is compliant in this area.  It's the decoder that's non-compliant by rejecting these long '0' bits, even though it works with other NMRA-compliant decoders!

**Comment by su68es** [comment]
Hello all,

that sounds good and I also had something like that as a thought.
I will look at the whole thing the next few days (will take a while, since the vacation is over).
I just do not know where the percentages come from. In the RCN-210 no relative values are specified but absolute:
Half of a "1" bit nominally: 58µs
Half of a "1" bit minimum at the decoder input: 52µs
Half of a "1" bit maximum at the decoder input: 64µs
Half of a "0" bit nominally: 100 µs
Half of a "0" bit minimum at decoder input: 90µs
Half of a "0"-bit maximum (if also an analog locomotive is to be controlled) at the decoder input: 10000µs
Half of a "0"-bit maximum at the decoder input: 119µs
But the exact values can be easily adjusted later.

Greetings
Roland

**Comment by Neil-McK** [comment]
The percentages are just what I calculated from the absolute range, so as
to be able to calculate a consistent margin for the 116us value.

**Comment by FrightRisk** [comment]
I agree with the idea for 4 modes. Are we voting? ;) Great ideas and a very
thorough analysis that I would like to edit and put as a technical document
on our we site.

Fred

**Comment by Neil-McK** [comment]
Feel free to share!

**Comment by su68es** [comment]
Hello all,

so, I have a new version ready and tested so far.
Before I upload this to Git I ask for a test and comments.
The files are available here:
https://1drv.ms/u/s!AiBvpacd_dyoipMUFMaMntoVy4kquw?e=tPzY0Q
Thanks in advance and greetings
Roland

**Comment by Neil-McK** [comment]
Roland,

Thanks for sending this through, it looks very good work.  I've had a go 
at testing it with a known command station, and also I've produced a 
sort of simulator for DCC signals that has configurable pulse lengths, 
adjustable prologue lengths etc.  and just transmits a choice of 
hard-coded DCC frames continuously.   From inspecting and running the 
code so far, I have the following comments:

1) I see that when the sample rate is below 2M samples/sec in the 
compliance modes, PulseView displays "The decoder reported an error".  
It's only by looking in the log that one can see what the error is.  I 
envisaged that the user would be able to see the actual error message in 
the Error/Info row of the decoder output similar to below - I realise I 
was a bit vague about that.  Is it possible to do this?

2) I also see that a bit length that is slightly outside of the margins 
in question still causes the entire packet to be rejected and a 
resynchronisation to be forced (see picture below).
As I said before, this restricts the available information as one can't 
see the context (i.e. what sort of DCC packet) where the bit error is 
occurring.  Ideally I would like for a message to appear in the 
Error/Info row highlighting the suspect bit(s), but the DCC bits and 
frame rows still show the packet contents, along the lines of the 
following composite image but perhaps saying 'Pulse length out of 
specification' or similar:
Would this be too complicated to implement?  If so, I could live with it 
as it is, by switching between timing modes, using the strict mode to 
see the out-of-spec bit lengths, and the more relaxed mode to see the 
packet contents.

3) Although I'm mainly interested in using the NMRA timings, I think 
that the RCN timings should be revised as they don't really fit with the 
RCN-210 specification at the moment, being more strict than the spec, 
rather than allowing for manufacturing tolerances.   RCN-210 specifies 
for command station compliance, the '1' bit should be generated as 58us 
+/- 3us, i.e. 55-61 us (a tolerance of 5%).    Likewise, the zero bit 
should be generated as 100us, allowing 95us (-5%) up to 9900us or 116us, 
with no additional tolerance, if read literally.  Since the most likely 
reason for allowing 116us would be to simplify the pulse generation ('0' 
bit is twice '1' bit), I strongly believe that a 5% tolerance should be 
allowed on the positive side too, i.e. a '0' pulse length maximum of 
122us for RCN compliance for controllers.  For decoders (the more 
relaxed RCN mode) the upper bound would be 116+10% = 128us.  In any 
case, the values of 97 and 114 currently used are tighter than the RCN 
requires, so bits of 96us or 116us are rejected that are RCN-compliant!  
  I've highlighted the values which I believe should be changed.

         timing = [  [0,  0,  0,  0,  0,     0],      #invalid
                     [52, 64, 6,  90, 10000, 10000],  #NMRA decoding
                     [52, 64, 6,  90, 128,   10000],  #RCN decoding
                     [55, 61, 3,  95, 9900,  9900],   #NMRA compilance 
testing
                     [55, 61, 3,  95, 122,   9898],   #RCN compilance 
testing
                     [0,  0,  0,  0,  0,     0]    ]  #Experimental

4) I don't have a Railcom system and I'm not likely to, but I have tried 
producing a Railcom cut-out simulator which goes to zero volts for 500us 
on the second half of a DCC '0' sent between idle packets.  Am I 
properly understanding how Railcom cutouts are produced?   Here's the 
logic analyser output, which looks fairly reasonable and accepts the 
railcom cutout as a stretched zero bit, although it doesn't specifically 
recognise it as a Railcom cutout.
If however the Railcom cutout is preceded by a '1' half-bit then the 
decoder loses synchronisation and doesn't resync until it finds a packet 
not preceded by a cutout (see below).  Whilst a '1' half-bit followed by 
a '0' half-bit isn't permitted within a DCC packet, it is not within the 
packet here so the prologue bits that follow it should be enough to 
resync bfor the next packet after the cutout.  I don't know whether such 
a cutout is valid within the spec or not, but it shouldn't cause the 
following packet to be rejected in any case.
The Railcom tests were performed with the timing mode set to "NMRA", so 
is not particularly strict.  Note that RCN decoders will have to accept 
Railcom presumably, so 500us has to be accepted by a decoder as a valid 
DCC zero half-bit?

Does this give some food for thought?

Best regards,

Neil

------ Original Message ------

**Comment by Neil-McK** [comment]
_[Images didn't appear so I've reposted direct to the site.  And I've done some further work on Railcom and revised comment 4 accordingly.  Then I found that the images didn't paste correctly!  It seems about right now.]_

Roland,

Thanks for sending this through, it looks very good work.  I've had a go at testing it with a known command station, and also I've produced a sort of simulator for DCC signals that has configurable pulse lengths, adjustable prologue lengths etc.  and just transmits a choice of hard-coded DCC frames continuously.   From inspecting and running the code so far, I have the following comments:

1) I see that when the sample rate is below 2M samples/sec in the compliance modes, PulseView displays "The decoder reported an error".  It's only by looking in the log that one can see what the error is.  I envisaged that the user would be able to see the actual error message in the Error/Info row of the decoder output similar to below - I realise I was a bit vague about that.  Is it possible to do this?
![image](https://user-images.githubusercontent.com/75813993/104239871-1d911c00-5453-11eb-9a65-1b65f6ed4f5d.png)


2) I also see that a bit length that is slightly outside of the margins in question still causes the entire packet to be rejected and a resynchronisation to be forced (see picture below). 
![image](https://user-images.githubusercontent.com/75813993/104244353-29341100-545a-11eb-8480-8736cd558b7a.png)
As I said before, this restricts the available information as one can't see the context (i.e. what sort of DCC packet) where the bit error is occurring.  Ideally I would like for a message to appear in the Error/Info row highlighting the suspect bit(s), but the DCC bits and frame rows still show the packet contents, along the lines of the following composite image but perhaps saying 'Pulse length out of specification' or similar:
![image](https://user-images.githubusercontent.com/75813993/104244389-36e99680-545a-11eb-8398-e389c7caf3f0.png)
Would this be too complicated to implement?  If so, I could live with it as it is, by switching between timing modes, using the strict mode to see the out-of-spec bit lengths, and the more relaxed mode to see the packet contents.

3) Although I'm mainly interested in using the NMRA timings, I think that the RCN timings should be revised as they don't really fit with the RCN-210 specification at the moment, being more strict than the spec, rather than allowing for manufacturing tolerances.   RCN-210 specifies for command station compliance, the '1' bit should be generated as 58us +/- 3us, i.e. 55-61 us (a tolerance of 5%).    Likewise, the zero bit should be generated as 100us, allowing 95us (-5%) up to 9900us or 116us, with no additional tolerance, if read literally.  Since the most likely reason for allowing 116us would be to simplify the pulse generation ('0' bit is twice '1' bit), I strongly believe that a 5% tolerance should be allowed on the positive side too, i.e. a '0' pulse length maximum of 122us for RCN compliance for controllers.  For decoders (the more relaxed RCN mode) the upper bound would be 116+10% = 128us.  In any case, the values of 97 and 114 currently used are tighter than the RCN requires, so bits of 96us or 116us are rejected that are RCN-compliant!   I've highlighted the values which I believe should be changed. 

        timing = [  [0,  0,  0,  0,  0,     0],      #invalid
                    [52, 64, 6,  90, 10000, 10000],  #NMRA decoding
                    [52, 64, 6,  90, 128,   10000],  #RCN decoding
                    [55, 61, 3,  95, 9900,  9900],   #NMRA compilance testing
                    [55, 61, 3,  95, 122,   9898],   #RCN compilance testing
                    [0,  0,  0,  0,  0,     0]    ]  #Experimental

4) I realise that the recent changes are nothing to do with Railcom, and you may not be supporting Railcom, but I thought it was worth testing what happens to the decoder when it encounters a Railcom cutout in the DCC signal.   I don't have a Railcom system and I'm not likely to soon, but I have tried producing a Railcom cut-out simulator which goes to zero volts for 500us after the stop bit of the packet, then follows it by a 100us half-bit before sending the next preamble.  Am I properly understanding how Railcom cutouts are produced?   Here's the logic analyser output, which looks fairly reasonable and accepts the railcom cutout as a stretched zero bit, although it doesn't specifically recognise it as a Railcom cutout.
![image](https://user-images.githubusercontent.com/75813993/104241507-b32dab00-5455-11eb-9369-62aa6d10612e.png)

However, if I don't have the 100us pulse, then the decoder misses the following packet.

The Railcom tests were performed with the timing mode set to "NMRA".

Does this give some food for thought? 

Best regards,

Neil

**Comment by Neil-McK** [comment]
I should fully complete my research before posting.  On the Railcom topic, I based my comment on one source on the web, but the NMRA Spec S-9.3.2 2012.12.10 doesn't require a zero half-bit between the cutout and the next preamble, in fact the diagram there shows the cutout overlaid on a stream of 1 bits going into the preamble:
![image](https://user-images.githubusercontent.com/75813993/104242649-60ed8980-5457-11eb-9617-15cd7af517df.png)
When I try this in the simulator, the decoder does not get the correct edge detection until the preamble has completed, so the packet following the cutout is rejected:
![image](https://user-images.githubusercontent.com/75813993/104243475-84650400-5458-11eb-9f70-cf2d7b207bb8.png)

The decoder should be capable of identifying the correct edge when it receives the first zero half-bit that follows the preamble and then decoding the packet from there.

Here's a better picture showing one packet decoded, followed by a cutout, then the next packet is not decoded:
![image](https://user-images.githubusercontent.com/75813993/104244102-af038c80-5459-11eb-85c7-33a5ffa0a560.png)

Neil.

**Comment by su68es** [comment]
Hello Neil,

this is fantastic. Thank you so much for your work.
I will take a closer look and check the individual points.
This week and the weekend I will (unfortunately) have little time to do so.
After a first quick look through, this is how I judge it:
1. I think that sigrik can not do otherwise.
2. I understand the point. During testing I managed several times after a timing error to recognize the subsequent telegrams as valid, which was not correct. Hence the strict resynchronization. But I think I know what you're getting at and need to try something (additional feature in 'expert mode').
3. I have to check it out calmly.
4. Something is wrong here. I need some time to check it. The automatic phase detection is quite tricky and I am not sure if it can always work. Therefore also now the possibility to select the phase manually.

Is it possible for you to make the *.sr files available to me so that I can use them for testing?
Thanks again and please be patient.
Roland

**Comment by Neil-McK** [comment]
I've attached a couple of sr files, one with the simulated Railcom cutout and one which was produced from DCC++ controller.  It only contains Idle frames, but it does have a few glitches which are a feature of the DCC++ex Isince its pulses are generated by digitalWrite from the interrupt handler). 

The Railcom simulation carries a health warning that it's my best attempt to simulate what's in the NMEA specs, but may be wrong.  I can easily tailor it if you want to try a different profile.

On point 1, I assume it is possible to put an annotation to the Error/Warning row that spans the entire list of sample data from first to last?  Wouldn't that do it?

On point 3, I apologise for going on about it.  Of course I can set the values through the interface, so you must use what you think is right irrespective of what I desire.  I don't intend to annoy you.

On point 4, in my decoder I look out for a mismatch between two adjacent half-bits.  I treat the preamble as a sequence of short half bits which may be an even or odd number of identical half-bits; when enough short half-bits (58us) are followed by a long half-bit (>=90us) then the long half-bit is treated as the first half of the start bit, and the decoder is now synchronised with the bitstream (whatever the polarity).  Then the decoder will read the packet successfully. 

[Sigrok captures.zip](https://github.com/littleyoda/sigrok-DCC-Protocoll/files/5823437/Sigrok.captures.zip)

**Comment by su68es** [comment]
Hello,

I have provided a new version:

[Link](https://1drv.ms/u/s!AiBvpacd_dyoipMUFMaMntoVy4kquw?e=q0Qnvo)

To the points:
1. True, if you do not try to throw an error over raise, you can use the normal display. Is now implemented so.

2. Is also implemented. In the Experimental settings select a generous timing and then set 'compare timing' to 'on'.

3. I welcome the discussion. I find the descriptions here ambiguous. Therefore, I have taken the values from the table that summarizes the technical data. In fact, there are three different value tuples here:
a) Signal at the output of the control panel
b) Signal at the track
c) signal which a decoder must accept
I have included the 3rd tuple resulting in the following timing:
 timing = [  [0,  0,  0,  0,  0,     0],      #invalid
                [52, 64, 6,  90, 10000, 10000],  #NMRA decoding
                [52, 64, 6,  90, 119,   10000],  #RCN decoding
                [55, 61, 3,  95, 9900,  9900],   #NMRA compliance testing
                [55, 61, 3,  95, 116,   9900],   #RCN compliance track
                [56, 60, 3,  97, 114,   9898],   #RCN compliance station
                [0,  0,  0,  0,  0,     0]    ]  #Experimental
![image](https://user-images.githubusercontent.com/76778984/105707322-0baf7e80-5f13-11eb-8c2f-77ec08385482.png)

4. I have changed the complete timing with the detection of a signal and now also use the search for a valid packet as the specification provides or as suggested by Neil. Thanks for the hint.

Overall, a major rebuild in the timing of the signal evaluation. I could not test everything yet. If there are any comments or wishes, they are welcome. Otherwise I'm still testing something and will upload the version then in Git.
Thanks to all for the input.

Roland

**Comment by Neil-McK** [comment]
Hi Roland,

Sorry I haven't been back to you, family illness has taken my time.  I hope
to look at this in the next few days, but your responses look very positive.

Neil

**Comment by su68es** [comment]
Hi Neil,

thanks for the information.
First of all all the best for the family.
There is no hurry for the review. I'm glad you're taking the time at all.
It's all just a (very nice) hobby.

Greetings
Roland

**Comment by Neil-McK** [comment]
Hi Roland,

I've just realised it's nearly a month and a half and I haven't been back to you.  We've had one family funeral in February, and another coming up next week.  Neither are related to COVID.  

I've put a few samples through the decoder version you linked above and I'm happy with it from my perspective. 

Best regards,

Neil.

**Comment by su68es** [comment]
Hi Neil,

thanks for the feedback and the tests.
I will test some more and then upload the version.
All the best and greetings

Roland

**Comment by FrightRisk** [comment]
@su68es Roland, it looks like you have another fork where you did work on RCN. Do you intend to issue a PR for us to review to include this in this main repository?

**Comment by su68es** [comment]
Hi Neil,
right, I didn't want to keep saddling littleyoda with work. He already wanted to push the project to sigrok and I think the repository there is the best place.
Currently the latest sources are here:
https://github.com/su68es/sigrok-DCC-Protocoll.
I'm trying to move the project to sigrok again. Let's see if this works.
Thanks and greetings
Roland

**Comment by su68es** [comment]
Hi Neil,
I just saw you responded in sigrok.
Thank you very much for your support. I can't quite get to grips with Gerhard's requirements. Let's see how it goes on.
Thanks again and greetings
Roland

**Comment by littleyoda** [comment]
FYI ...I send the following mail to the sigrok-mailinglist:

> Hi all,
> short feedback from my side (littleyoda)
> unfortunately I never had the time/nerve to really push my implementation into sigrok.
> In the meantime su68es had done a great job, in developing a better implementation.
>Ignores my repository and take his.
>Gruß,
> Sven

**Comment by FrightRisk** [comment]
So we (Roland) will try to push it to Sigrok and we maintain it from there? Works for me. If you need help, I can find time in the next few days to struggle through Sigrok. I know I can help with a Wiki

**Comment by su68es** [comment]
I think this is a good solution to store the decoder in sigrok.
Many of Gerhard's conditions are already fulfilled.
There are already example captures
https://sigrok.org/gitweb/?p=sigrok-dumps.git;a=tree;f=dcc;h=9860ae8c08f4b7dba2dabc1bd1016446fa952cb4;hb=HEAD
and in my repository the decoder is available and further examples together with a readme.
I will work on the unit tests in the next days (never done before) as time allows.
Help regarding the WIKI is welcome.

**Comment by su68es** [comment]
Hi Neil, Sven,
I am not getting anywhere with sigrok.
I will continue to maintain the decoder in my repository.
https://github.com/su68es/sigrok-DCC-Protocoll
Thanks for your support.
Greetings
Roland
